sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,133 @@
+
+#### Issue
+
+Endpoints object for a service contains all the individual endpoints of that service. As a result, whenever even a single pod in a service is added/updated/deleted, the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers. Not being able to efficiently read/update individual endpoint changes can lead to (for e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If you consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+Overall, this is a serious performance drawback affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). The current endpoints API (which was designed at a very early stage of kubernetes when people weren't really thinking too much about scalability/performance) makes it hard to solve this problem without introducing breaking changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API with (at least) the following properties:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Separate API object for each individual endpoint
+  - Has a 'Ready' property as part of the object
+  - Port ranges and other features we’ve put off for a long time
+
+
+## Define/measure performance SLIs for endpoints propagation
+
+#### Issue
+
+We need to measure the e2e endpoints propagation latency, i.e the time from when a 
+service is created until its endpoints are populated, given that all of the associated pods 
+are already running and ready. We also need to come up with reasonable SLOs for the 
+same and verify that they’re satisfied at scale.
+
+#### Relevant issues/links
+
+- [#10436](https://github.com/kubernetes/kubernetes/issues/10436)
+- This [doc on networking SLIs](https://docs.google.com/document/d/1kudI8uILO13ySkpE5HgNHvIGn-yXVQatpyM3YTVVUis) proposes one way of formulating the SLI
+
+#### Possible Solution(s)
+
+- Measure the SLI defined in the above doc and come up with a reasonable SLO
+
+
+## Master election shouldn’t consume unnecessary watch bandwidth
+
+#### Issue
+
+Currently we store the leader lock for master components like the scheduler in the corresponding endpoints object. This has the undesirable side effect of sending a notification down the kube-proxy <-> master (and kube-dns <-> master) watch every second, which is not actually required.
```

Comment:
Why haven’t we moved to configmaps?

Reply:
I guess noone really picked this up (we would need to figure our migration in HA setups, which is doable but requires some work).

Reply:
Most of the code should have switched.  Scheduler and CM should both be on config map.  Some of the ecosystem listeners have moved.  OpenShift moved in 3.9 / 1.9 to force all users to use config map.

Reply:
I agree that they should have switched. But given that they didn't, I think we shouldn't do it now and instead wait for Lease object (the API is already merged in fact) and switch things directly to that (and deprecate both endpoint and configmap for that purpose).